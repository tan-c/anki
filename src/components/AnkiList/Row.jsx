import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { Map } from 'immutable';
import { Grid, Icon } from 'semantic-ui-react';

import { AnkiActions, activeAnkiSelector } from 'utility-redux/anki';
import { UiActions } from 'utility-redux/ui';

export class AnkiRow extends React.Component {
  render() {
    const { anki, activeAnki } = this.props;

    return (
      <Grid.Row
        style={{
          padding: 0,
          minHeight: 30
        }}
      >
        <Grid.Column
          width={1}
        >
          {`${anki.getIn(['revision', 'round'])}/${anki.getIn(['revision', 'passing'])} - `}
        </Grid.Column>

        <Grid.Column
          width={12}
          textAlign="left"
        >
          {anki.get('question')}
        </Grid.Column>

        <Grid.Column
          width={3}
          textAlign="center"
        >
          <span style={{
            width: 20
          }}
          >
            {anki.has('tags') ? anki.get('tags').size : 0}
          </span>

          <Icon
            name="close"
            onClick={_ => this.props.AnkiActions.deleteRecord(anki)}
          />

          <Icon
            name="edit"
            onClick={_ => this.props.UiActions.updateIn(['activeAnkiId'], anki.get('_id'))
            }
          />
        </Grid.Column>
      </Grid.Row>
    );
  }
}

AnkiRow.defaultProps = {
  anki: Map(),
  activeAnki: Map()
};

AnkiRow.propTypes = {
  anki: PropTypes.object,
  activeAnki: PropTypes.object,

  AnkiActions: PropTypes.object.isRequired,
  UiActions: PropTypes.object.isRequired
};

function mapStateToProps(state, ownProps) {
  return {
    anki: ownProps.anki,
    activeAnki: activeAnkiSelector(state)
  };
}

function mapDispatchToProps(dispatch) {
  return {
    AnkiActions: bindActionCreators(AnkiActions, dispatch),
    UiActions: bindActionCreators(UiActions, dispatch)
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AnkiRow);
