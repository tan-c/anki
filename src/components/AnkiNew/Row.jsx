import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { Map } from 'immutable';
import { Grid, Icon } from 'semantic-ui-react';

import { AnkiActions, activeAnkiSelector } from 'utility-redux/anki/anki';
import { UiActions } from 'utility-redux/common/ui';

export class AnkiRow extends React.Component {
  render() {
    const { anki, activeAnki } = this.props;

    return (
      <Grid.Row style={{
        padding: 0,
        minHeight: 30
      }}
      >
        <div style={{
          width: 30
        }}
        >
          {`${anki.getIn(['revision', 'round'])}/${anki.getIn(['revision', 'passing'])} - `}
        </div>

        <Grid.Column>
          {anki.get('question')}
        </Grid.Column>

        <div style={{
          width: 20
        }}
        >
          {anki.has('tags') ? anki.get('tags').size : 0}
        </div>

        <Icon
          name="close"
          onClick={_ => this.props.AnkiActions.deleteRecord(anki)}
        />

        <Icon
          name="edit"
          onClick={_ => this.props.UiActions.updateIn(['activeAnkiId'], anki.get('_id'))
          }
        />
      </Grid.Row>
    );
  }
}

AnkiRow.defaultProps = {
  anki: Map(),
  activeAnki: Map()
};

AnkiRow.propTypes = {
  anki: PropTypes.object,
  activeAnki: PropTypes.object,

  AnkiActions: PropTypes.object.isRequired,
  UiActions: PropTypes.object.isRequired
};

function mapStateToProps(state, ownProps) {
  return {
    anki: ownProps.anki,
    activeAnki: activeAnkiSelector(state)
  };
}

function mapDispatchToProps(dispatch) {
  return {
    AnkiActions: bindActionCreators(AnkiActions, dispatch),
    UiActions: bindActionCreators(UiActions, dispatch)
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AnkiRow);
