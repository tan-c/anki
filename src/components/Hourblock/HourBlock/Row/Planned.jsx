import React from 'react';
import PropTypes from 'prop-types';
import { Map } from 'immutable';
import { connect } from 'react-redux';
import { currentUserSelector } from 'utility-redux/user';
import ProjectSelectConnected from 'utility-react-component/Form/HourblockProjectSelect';

// import { projectFirstTaskSelector } from 'utility-redux/task';

// FIXME: should be using the input controlled here but keydown event is different
// import InputControlled from 'utility-react-component/Form/Input/Controlled';

export class HourBlockRowPlanned extends React.Component {
  constructor(props, context) {
    super(props, context);
    this.state = {
      mainTask: props.plannedPomo.hasIn(['tasks', 'main']) ? props.plannedPomo.getIn(['tasks', 'main']) : '',
      minorTask: props.plannedPomo.hasIn(['tasks', 'minor']) ? props.plannedPomo.getIn(['tasks', 'minor']) : '',
      // recurTask: '',
    };
  }

  componentWillReceiveProps(nextProps) {
    ['main', 'minor'].forEach((t) => {
      if (nextProps.plannedPomo.hasIn(['tasks', t]) && nextProps.plannedPomo.getIn(['tasks', t]) !== this.state[`${t}Task`]) {
        this.setState({
          [`${t}Task`]: nextProps.plannedPomo.getIn(['tasks', t]),
        });
      }
    });
  }

  shouldComponentUpdate(nextProps, nextState) {
    let shouldUpdate = false;
    Object.keys(nextProps).forEach((key) => {
      if (JSON.stringify(this.props[key]) !== JSON.stringify(nextProps[key])) {
        shouldUpdate = true;
      }
    });

    Object.keys(nextState).forEach((key) => {
      if (JSON.stringify(this.state[key]) !== JSON.stringify(nextProps[key])) {
        shouldUpdate = true;
      }
    });

    return shouldUpdate;
  }

  renderTasks = (isTodayPast, taskFromPomo) => (
    <React.Fragment>
      <span className="flex-3 border-right-white padding-horizontal-5 text-left">
        {!isTodayPast && taskFromPomo}
      </span>

      {/* {!window.isMobile && showMinorTask &&
        <span
          className={`${isTodayPast ? 'bg-white-10' : ''} flex-2 border-right-white padding-horizontal-5 text-left`}
        >
          {minorTask}
        </span>} */}
    </React.Fragment>
  )

  renderMainTaskInput = (mainTask, isTodayPast) => {
    const {
      sectionOfDay, onChangePlannedPomo, plannedPomo,
      isUpdatingPlannedPomo, recordPomo
    } = this.props;

    if (recordPomo.has('content')) {
      return (
        <span className="text-left flex-3 border-right-white-20 padding-left-5">
          {recordPomo.get('content')}
        </span>
      );
    }

    return (
      <input
        type="text"
        className={`flex-3 border-right-white-20 ${isTodayPast && 'bg-black'}`}
        name="tasks.main"
        ref={(ref) => { this.hourblockPlanneMainInput = ref; }}
        value={mainTask}
        disabled={isUpdatingPlannedPomo || isTodayPast}
        onChange={(event) => {
          this.setState({
            mainTask: event.target.value,
          });
        }}
        onKeyDown={(event) => {
          if (event.keyCode === 13) {
            onChangePlannedPomo(sectionOfDay, plannedPomo, event);
            this.hourblockPlanneMainInput.value = '';
          }
        }}
      />
    );
  }

  renderMinorTaskInput = (minorTask, isTodayPast) => {
    const {
      sectionOfDay, onChangePlannedPomo, plannedPomo,
      isUpdatingPlannedPomo
    } = this.props;

    return (
      <input
        type="text"
        className={`flex-2 border-right-white-20 ${isTodayPast && 'bg-black'}`}
        name="tasks.minor"
        ref={(ref) => { this.hourblockPlanneMinorInput = ref; }}
        value={minorTask}
        disabled={isUpdatingPlannedPomo}
        onChange={(event) => {
          this.setState({
            minorTask: event.target.value,
          });
        }}
        onKeyDown={(event) => {
          if (event.keyCode === 13) {
            onChangePlannedPomo(sectionOfDay, plannedPomo, event);
            this.hourblockPlanneMinorInput.value = '';
          }
        }}
      />
    );
  }

  renderRecurTaskInput = (isTodayPast) => {
    const {
      sectionOfDay, onChangePlannedPomo, plannedPomo, currentUser,
    } = this.props;

    const enableChangeRecur = currentUser.hasIn(['config', 'enableChangeRecur']) && currentUser.getIn(['config', 'enableChangeRecur']);

    return (
      <input
        type="text"
        className={`${isTodayPast ? 'bg-black' : ''} ${window.isMobile ? 'flex-1' : 'width-200'} border-right-white`}
        name="tasks.recur"
        disabled={!enableChangeRecur || isTodayPast}
        ref={(ref) => { this.hourblockPlannedRecurInput = ref; }}
        placeholder={plannedPomo.getIn(['tasks', 'recur'])}
        onKeyDown={(event) => {
          if (event.keyCode === 13) {
            onChangePlannedPomo(sectionOfDay, plannedPomo, event);
            this.hourblockPlannedRecurInput.value = '';
          }
        }}
      />
    );
  }

  render() {
    const { mainTask, minorTask } = this.state;
    const {
      sectionOfDay, onChangePlannedPomo, plannedPomo, sectionName,
      currentUser,
      currentSectionOfDay,
      isToday,
      allProjectTasksOrdered,
      recordPomo,
      // projectFirstTask,
    } = this.props;

    // const mainTaskLocked = plannedPomo.has('mainTaskLocked') && plannedPomo.get('mainTaskLocked');
    const isTodayPast = isToday && currentSectionOfDay > sectionOfDay;
    const disableMainTaskInput = isToday && currentSectionOfDay > sectionOfDay;

    const showMinorTask = currentUser.hasIn(['config', 'showMinorTask']) && currentUser.getIn(['config', 'showMinorTask']);
    const currentPlannedPomoTask = allProjectTasksOrdered.hasIn([plannedPomo.getIn(['project', '_id']), 0]) ? allProjectTasksOrdered.getIn([plannedPomo.getIn(['project', '_id']), 0]) : Map();

    const taskContent = currentPlannedPomoTask.has('taskName') ? currentPlannedPomoTask.get('taskName') : '';
    // const isUnFinished = !recordPomo.hasIn(['project', '_id']) && isTodayPast;

    return (
      <React.Fragment>
        {window.isMobile
          && (
            <div className="flex-container flex-1">
              <div className="flex-container-row flex-1">
                <span className="border-right-white height-lineheight-30 width-50">
                  {sectionName}
                </span>

                {disableMainTaskInput === true && this.renderTasks(isTodayPast, taskContent, minorTask)}

                {disableMainTaskInput !== true && this.renderMainTaskInput(mainTask, isTodayPast)}
              </div>

              <div className="flex-container-row flex-1">
                <ProjectSelectConnected
                  className="width-30 border-right-white"
                  value={plannedPomo.getIn(['project', '_id'])}
                  onChangeEvent={event => onChangePlannedPomo(sectionOfDay, plannedPomo, event)}
                  color={plannedPomo.getIn(['project', 'category', 'color'])}
                />
                {this.renderRecurTaskInput(isTodayPast)}
              </div>
            </div>
          )
        }

        {!window.isMobile
          && (
            <React.Fragment>
              <span className="border-right-white width-50 height-lineheight-30">
                {sectionName}
              </span>

              {!isTodayPast && (
                <ProjectSelectConnected
                  className="width-80 border-right-white height-lineheight-30"
                  value={plannedPomo.getIn(['project', '_id'])}
                  onChangeEvent={event => onChangePlannedPomo(sectionOfDay, plannedPomo, event)}
                  color={plannedPomo.getIn(['project', 'category', 'color'])}
                />
              )}

              {isTodayPast
              && (
                <span className="width-80 border-right-white height-lineheight-30" style={{ backgroundColor: plannedPomo.getIn(['project', 'category', 'color']) }}>
                  {plannedPomo.getIn(['project', 'name'])}
                </span>
              )
              }

              <div className="flex-1 flex-container-row">
                {/* {this.renderTasks(isTodayPast, taskContent)} */}
                {this.renderMainTaskInput(mainTask, isTodayPast)}
                {showMinorTask && this.renderMinorTaskInput(minorTask, isTodayPast)}
                {this.renderRecurTaskInput(isTodayPast)}
              </div>
            </React.Fragment>
          )
        }
      </React.Fragment>
    );
  }
}

HourBlockRowPlanned.defaultProps = {
  currentUser: Map(),
  currentSectionOfDay: 0,
  sectionOfDay: 0,
  sectionName: '',
  plannedPomo: Map(),
  recordPomo: Map(),
  onChangePlannedPomo: () => { },
  // projectFirstTask: Map(),
};

HourBlockRowPlanned.propTypes = {
  isToday: PropTypes.bool.isRequired,
  currentSectionOfDay: PropTypes.number,
  currentUser: PropTypes.object,
  sectionOfDay: PropTypes.number,
  sectionName: PropTypes.string,
  plannedPomo: PropTypes.object,
  recordPomo: PropTypes.object,
  onChangePlannedPomo: PropTypes.func,
  isUpdatingPlannedPomo: PropTypes.bool.isRequired,
  allProjectTasksOrdered: PropTypes.object.isRequired,
  // projectFirstTask: PropTypes.object,
};

function mapStateToProps(state, ownProps) {
  return {
    isToday: ownProps.isToday,
    currentSectionOfDay: ownProps.currentSectionOfDay,
    sectionOfDay: ownProps.sectionOfDay,
    sectionName: ownProps.sectionName,
    plannedPomo: ownProps.plannedPomo,
    recordPomo: ownProps.recordPomo,
    onChangePlannedPomo: ownProps.onChangePlannedPomo,
    allProjectTasksOrdered: ownProps.allProjectTasksOrdered,
    currentUser: currentUserSelector(state),
    isUpdatingPlannedPomo: ownProps.isUpdatingPlannedPomo,
    // projectFirstTask: projectFirstTaskSelector(state, ownProps),
  };
}

export default connect(mapStateToProps)(HourBlockRowPlanned);
