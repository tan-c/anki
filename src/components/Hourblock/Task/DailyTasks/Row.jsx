import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Map, isImmutable } from 'immutable';
import moment from 'moment-timezone';

import {
  TaskActions,
  dailyTasksSelector,
} from 'utility-redux/hourblock/task';

import Input from 'utility-react-component/Form/Input/Uncontrolled';
import DefaultSelect from 'utility-react-component/Form/Select';
import InputNewConnected from 'utility-react-component/Form/Input/New';

class DailyTasksRow extends React.Component {
  render() {
    const { dailyTasks, selectedYearlyTask, dayVal } = this.props;

    return (
      <div className={`${dayVal.dayMomentObject.unix() === moment().tz('Asia/Tokyo').startOf('day').unix() && 'border-top-orange'} ${dayVal.dayMomentObject.isoWeekday() === 7 && 'margin-bottom-10'}`}>
        <div className="flex-container-row typical-setup">
          <span className={`flex-1 ${dayVal.dayMomentObject.isoWeekday() > 5 && 'color-orange'}`}>
            {dayVal.dayMomentObject.format('MM-DD')}
          </span>

          {dayVal.dayMomentObject.unix() >= moment().tz('Asia/Tokyo').startOf('day').unix()
          && (
            <InputNewConnected
              inputName="content"
              inputClassNames="flex-3"
              newRecord={{
                targetCompletion: dayVal.dayMomentObject,
                type: 'daily',
                project: selectedYearlyTask.getIn(['project', '_id']),
              }}
              actions={this.props.TaskActions}
            />
          )}

          {dayVal.dayMomentObject.unix() < moment().tz('Asia/Tokyo').startOf('day').unix()
          && <span className="flex-3" />
          }

          <span className="width-20">
            {dailyTasks.has(dayVal.dayOfYearString)
            && (
              <div>
                {dailyTasks.get(dayVal.dayOfYearString).size === 1 && dailyTasks.getIn([dayVal.dayOfYearString, '0', 'estimateHour'])}
                {dailyTasks.get(dayVal.dayOfYearString).size > 1
              && dailyTasks.get(dayVal.dayOfYearString).reduce((currentSum, b) => (isImmutable(currentSum) ? currentSum.get('estimateHour') : currentSum) + b.get('estimateHour'))}
              </div>
            )}
          </span>
        </div>

        {dailyTasks.has(dayVal.dayOfYearString) && dailyTasks.get(dayVal.dayOfYearString).map(task => (
          <React.Fragment key={task.get('_id')}>
            <div
              className={`${task.getIn(['project', '_id']) === selectedYearlyTask.getIn(['project', '_id']) && 'bg-orange'}  flex-container-row typical-setup border-bottom`}
              style={{ borderLeft: `5px solid ${task.hasIn(['project', 'category', 'color']) ? task.getIn(['project', 'category', 'color']) : 'white'}` }}
            >
              <Input
                inputName="content"
                inputClassNames="flex-4"
                record={task}
                actions={this.props.TaskActions}
              />

              <Input
                inputName="estimateHour"
                inputClassNames="flex-1"
                record={task}
                actions={this.props.TaskActions}
              />

              <DefaultSelect
                className="flex-1"
                onChangeEvent={(event) => {
                  this.props.TaskActions.update(task.set('recur', event.target.value));
                }}
                value={task.get('recur')}
                options={['none', 'daily', 'weekly', 'monthly', 'yearly']}
              />

              <i
                className="fa fa-fw fa-close width-15"
                role="button" tabIndex="-1"
                onClick={_ => this.props.TaskActions.deleteRecord(task)}
              />

              <i
                className="fa fa-fw fa-check width-15"
                role="button"
                tabIndex="-1"
                onClick={_ => this.props.TaskActions.update(task.setIn(['project', '_id'], selectedYearlyTask.getIn(['project', '_id'])))}
              />

              <i
                className="fa fa-fw fa-gift width-15"
                role="button"
                tabIndex="-1"
                onClick={_ => this.props.TaskActions.update(task.set('reward', ''))}
              />
            </div>

            {task.has('reward') && (
              <div className="flex-container-row typical-setup border-bottom-white border-left-white border-right-white">
                <Input
                  inputName="reward"
                  inputClassNames="flex-1"
                  record={task}
                  actions={this.props.TaskActions}
                />
              </div>
            )}
          </React.Fragment>

        ))}
      </div>
    );
  }
}

DailyTasksRow.defaultProps = {
  dailyTasks: Map(),
};

DailyTasksRow.propTypes = {
  dailyTasks: PropTypes.object,
  dayVal: PropTypes.object.isRequired,
  selectedYearlyTask: PropTypes.object.isRequired,

  TaskActions: PropTypes.object.isRequired,
};

function mapStateToProps(state, ownProps) {
  return {
    dailyTasks: dailyTasksSelector(state),
    selectedYearlyTask: ownProps.selectedYearlyTask,
    dayVal: ownProps.dayVal,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    TaskActions: bindActionCreators(TaskActions, dispatch),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(DailyTasksRow);
