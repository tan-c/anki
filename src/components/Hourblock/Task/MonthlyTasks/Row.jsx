import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { Map } from 'immutable';

import {
  TaskActions,
  selectedYearlyTaskSelector,
} from 'utility-redux/task';

import Input from 'utility-react-component/Form/Input/Uncontrolled';

export class MonthlyTasksRow extends React.Component {
  render() {
    const { task, selectedYearlyTask } = this.props;

    return (
      <React.Fragment>
        <div
          data-role="monthlytasks-row"
          className={`${task.getIn(['project', '_id']) === selectedYearlyTask.getIn(['project', '_id']) && 'bg-orange'} flex-container-row typical-setup border-bottom`}
          style={{ borderLeft: `5px solid ${task.hasIn(['project', 'category', 'color']) ? task.getIn(['project', 'category', 'color']) : 'white'}` }}
        >
          <Input
            inputName="content"
            inputClassNames="flex-3"
            record={task}
            actions={this.props.TaskActions}
          />

          <i
            className="fa fa-fw fa-close width-15"
            role="button" tabIndex="-1"
            onClick={_ => this.props.TaskActions.deleteRecord(task)}
          />

          <i
            className="fa fa-fw fa-check width-15"
            role="button"
            tabIndex="-1"
            onClick={_ => this.props.TaskActions.update(task.setIn(['project', '_id'], selectedYearlyTask.getIn(['project', '_id'])))}
          />

          <i
            className="fa fa-fw fa-gift width-15"
            role="button"
            tabIndex="-1"
            onClick={_ => this.props.TaskActions.update(task.set('reward', ''))}
          />
        </div>

        {task.has('reward') && (
          <div className="flex-container-row typical-setup border-bottom-white border-left-white border-right-white">
            <Input
              inputName="reward"
              inputClassNames="flex-1"
              record={task}
              actions={this.props.TaskActions}
            />
          </div>
        )}
      </React.Fragment>
    );
  }
}

MonthlyTasksRow.defaultProps = {
  selectedYearlyTask: Map(),
};

MonthlyTasksRow.propTypes = {
  task: PropTypes.object.isRequired,
  selectedYearlyTask: PropTypes.object,

  TaskActions: PropTypes.object.isRequired,
};

function mapStateToProps(state, ownProps) {
  return {
    task: ownProps.task,
    selectedYearlyTask: selectedYearlyTaskSelector(state),
  };
}

function mapDispatchToProps(dispatch) {
  return {
    TaskActions: bindActionCreators(TaskActions, dispatch),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(MonthlyTasksRow);
