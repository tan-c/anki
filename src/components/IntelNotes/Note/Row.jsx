import React from 'react';
import PropTypes from 'prop-types';
// import { Map } from 'immutable';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

// import toastr from 'toastr';

import { NoteActions } from 'utility-redux/note';
// import { UiActions } from 'utility-redux/ui';
import {
  UserActions,
  currentUserSelector,
  currentUserRecentNoteIdSelector,
  currentUserRecentNotebookIdSelector
} from 'utility-redux/user';

export class NoteRow extends React.Component {
  onDragOver = (event) => {
    // By default, data/elements cannot be dropped in other elements. To allow a drop, we must prevent the default handling of the element. This is done by calling the event.preventDefault() method for the ondragover attribute.

    // https:// www.w3schools.com/tags/ev_ondragover.asp
    event.preventDefault();
  }

  onDragStart = (event) => {
    const { note } = this.props;
    event.dataTransfer.setData('text', JSON.stringify({
      note,
      source: 'note',
    }));
  }

  onClickNote = () => {
    const { note, currentUser } = this.props;
    this.props.UserActions.update(currentUser.setIn(['config', 'hima', 'recentNote'], note.get('_id')));
  }

  updateDeletedFlagNote = (event) => {
    const { note } = this.props;
    event.stopPropagation();
    if (confirm('Really moving to dustbin?')) {
      this.props.NoteActions.update(note.set('isDeleted', true));
    }
  }

  deleteNote = (event) => {
    const { note } = this.props;
    event.stopPropagation();
    if (confirm('Really deleting?')) {
      this.props.NoteActions.deleteRecord(note);
    }
  }

  render() {
    const {
      note, activeNoteId, activeNotebookId,
    } = this.props;

    return (
      <div
        role="menuitem"
        tabIndex="-1"
        data-role="note-row"
        onDragOver={this.onDragOver}
        draggable="true"
        onDragStart={this.onDragStart}
        onClick={this.onClickNote}
        className={`flex-container-row typical-setup ${note.get('_id') === activeNoteId ? 'bg-red' : 'bg-orange-light'} 'border-right-orange-light-5px'`}
      >
        <span className="flex-1">
          {note.get('title')}
        </span>

        {activeNotebookId === 'deleted' && (
          <i
            role="button"
            tabIndex="-1"
            className="fa fa-fw fa-trash font-20"
            onClick={this.deleteNote}
          />
        )}

        {activeNotebookId !== 'deleted'
          && (
            <i
              role="button"
              tabIndex="-1"
              className="fa fa-fw fa-close font-20"
              onClick={this.updateDeletedFlagNote}
            />
          )}
      </div>
    );
  }
}

NoteRow.defaultProps = {
  activeNoteId: ''
};

NoteRow.propTypes = {
  currentUser: PropTypes.object.isRequired,
  note: PropTypes.object.isRequired,
  activeNoteId: PropTypes.string,
  activeNotebookId: PropTypes.string.isRequired,

  NoteActions: PropTypes.object.isRequired,
  UserActions: PropTypes.object.isRequired,
};

function mapStateToProps(state, ownProps) {
  return {
    note: ownProps.note,
    currentUser: currentUserSelector(state),
    activeNoteId: currentUserRecentNoteIdSelector(state),
    activeNotebookId: currentUserRecentNotebookIdSelector(state),
  };
}

function mapDispatchToProps(dispatch) {
  return {
    UserActions: bindActionCreators(UserActions, dispatch),
    NoteActions: bindActionCreators(NoteActions, dispatch),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(NoteRow);
