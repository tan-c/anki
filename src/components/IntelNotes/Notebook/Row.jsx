import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import toastr from 'toastr';
import { Map } from 'immutable';

import { NotebookActions } from 'utility-redux/anki/notebook';
import { NoteActions } from 'utility-redux/anki/note';
import { UiActions } from 'utility-redux/common/ui';
import InputUncontrolledConnected from 'utility-react-component/Form/Input/Uncontrolled';

export class NotebookRow extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      editMode: 0,
    };
  }

  shouldComponentUpdate(nextProps, nextState) {
    if (this.state.editMode !== nextState.editMode) {
      return true;
    }

    let shouldUpdate = false;
    Object.keys(nextProps).forEach((key) => {
      // If notes, update when created new notes or when deleting note
      if (key === 'notes') { // do not stringify as it is costly
        if (this.props[key].filter(note => note.get('isDeleted')).size === nextProps[key].filter(note => note.get('isDeleted')).size && this.props[key].count() !== nextProps[key].count()) {
          shouldUpdate = true;
        }
      } else if (JSON.stringify(this.props[key]) !== JSON.stringify(nextProps[key])) {
        // Note change should only trigger update if the number changed
        shouldUpdate = true;
      }
    });

    return shouldUpdate;
  }

  toggleEditMode = (event) => {
    event.stopPropagation(); // Do not trigger the click event to set activeNotebook
    // event.nativeEvent.stopImmediatePropagation();
    this.setState(prevState => ({
      editMode: 1 - prevState.editMode,
    }));
  }

  deleteNotebook = () => {
    if (confirm('Sure to delete this notebook')) {
      this.props.NotebookActions.deleteRecord(this.props.notebook);
    }
  }

  dragStart = (event, notebook) => {
    event.dataTransfer.setData('text', JSON.stringify({
      notebook,
      source: 'notebook',
    }));
  }

  dragOver = (event) => {
    event.preventDefault();
  }

  drop = (event) => {
    const {
      notebook,
    } = this.props;
    event.preventDefault();

    const data = JSON.parse(event.dataTransfer.getData('text'));
    if (data.source === 'note') {
      data.note.notebook = notebook.get('_id');
      this.props.NoteActions.update(data.note);
    } else {
      toastr.error('', 'Source Must Be Note');
    }
  }

  render() {
    const {
      notebook, activeNotebookId, notes,
    } = this.props;

    const { editMode } = this.state;

    return (
      <div
        data-role="notebook-row"
        role="menuitem" tabIndex="-1" draggable="true" onDragStart={event => this.dragStart(event, notebook)} onDragOver={this.dragOver} onDrop={this.drop} className={`flex-container-row typical-setup ${notebook.get('_id') === activeNotebookId && 'bg-red'}`} onClick={() => {
          this.props.UiActions.updateIn(['himalayan', 'activeNotebookId'], notebook.get('_id'));
        }}
      >
        <span data-test="notebook-title" className="flex-3">
          {!editMode ? notebook.get('title') : (
            <InputUncontrolledConnected
              inputName="title"
              record={notebook}
              actions={this.props.NotebookActions}
            />
          )}
        </span>

        { notes.filter(note => note.getIn(['notebook', '_id']) === notebook.get('_id')).count() === 0
          && <i role="menuitem" tabIndex="-1" className="fa fa-fw fa-close font-14" onClick={this.deleteNotebook} />
        }

        <i role="menuitem" tabIndex="-1" className="fa fa-fw fa-edit font-14" onClick={this.toggleEditMode} />

        <div className="width-15 bg-white-20 text-center">
          {notes.filter(note => note.getIn(['notebook', '_id']) === notebook.get('_id') && !note.get('isDeleted')).size}
        </div>
      </div>
    );
  }
}

NotebookRow.defaultProps = {
  notes: Map(),
  notebook: Map(),
  activeNotebookId: '',

  NotebookActions: {},
  NoteActions: {},
  UiActions: {},
};

NotebookRow.propTypes = {
  notes: PropTypes.object,
  notebook: PropTypes.object,
  activeNotebookId: PropTypes.string,

  NotebookActions: PropTypes.object,
  NoteActions: PropTypes.object,
  UiActions: PropTypes.object,
};

function mapStateToProps(state, ownProps) {
  return {
    notes: state.notes,
    notebook: ownProps.notebook,
    activeNotebookId: state.ui.getIn(['himalayan', 'activeNotebookId']),
  };
}

function mapDispatchToProps(dispatch) {
  return {
    NotebookActions: bindActionCreators(NotebookActions, dispatch),
    NoteActions: bindActionCreators(NoteActions, dispatch),
    UiActions: bindActionCreators(UiActions, dispatch),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(NotebookRow);
