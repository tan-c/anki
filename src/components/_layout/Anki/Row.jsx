import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { Map } from 'immutable';

import { AnkiActions, activeAnkiSelector } from 'utility-redux/anki/anki';
import { UiActions } from 'utility-redux/common/ui';

export class AnkiRow extends React.Component {
  render() {
    const { anki, activeAnki } = this.props;

    return (
      <div
        data-role="anki-row"
        className={`flex-container-row typical-setup ${
          activeAnki.get('_id') === anki.get('_id')
            ? 'bg-orange'
            : 'border-black-20'
        }`}
      >
        <span className="width-30 color-white text-center bg-green">
          {anki.getIn(['revision', 'round'])}
          {'/'}
          {anki.getIn(['revision', 'passing'])}
        </span>

        <span className="flex-1 margin-left-5">{anki.get('question')}</span>

        <span className="width-20 bg-orange text-center text-white">
          {anki.has('tags') ? anki.get('tags').size : 0}
        </span>
        <i
          role="button"
          tabIndex="-1"
          className="fa fa-close fa-fw font-14"
          onClick={_ => this.props.AnkiActions.deleteRecord(anki)}
        />

        <i
          role="button"
          tabIndex="-1"
          className="fa fa-fw fa-edit font-14"
          onClick={_ => this.props.UiActions.updateIn(['activeAnkiId'], anki.get('_id'))
          }
        />
      </div>
    );
  }
}

AnkiRow.defaultProps = {
  anki: Map(),
  activeAnki: Map()
};

AnkiRow.propTypes = {
  anki: PropTypes.object,
  activeAnki: PropTypes.object,

  AnkiActions: PropTypes.object.isRequired,
  UiActions: PropTypes.object.isRequired
};

function mapStateToProps(state, ownProps) {
  return {
    anki: ownProps.anki,
    activeAnki: activeAnkiSelector(state)
  };
}

function mapDispatchToProps(dispatch) {
  return {
    AnkiActions: bindActionCreators(AnkiActions, dispatch),
    UiActions: bindActionCreators(UiActions, dispatch)
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AnkiRow);
