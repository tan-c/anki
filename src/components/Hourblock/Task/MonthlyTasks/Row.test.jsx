import React from 'react';
import { shallow } from 'enzyme';
import expect from 'expect';
import toJson from 'enzyme-to-json';

import { Map } from 'immutable';

import { MonthlyTasksRow } from './Row';

function setup() {
  const props = {
    task: Map(),

    TaskActions: {},
  };

  return shallow(<MonthlyTasksRow {...props} />);
}

describe('MonthlyTasksRow Unit Tests', () => {
  describe('when normally set up', () => {
    const wrapper = setup();
    it('renders correctly', () => {
      const jsonOutput = toJson(wrapper);
      expect(jsonOutput).toMatchSnapshot();
      expect(wrapper.find("div[data-role='monthlytasks-row']")).toHaveLength(1);
    });
  });
});
