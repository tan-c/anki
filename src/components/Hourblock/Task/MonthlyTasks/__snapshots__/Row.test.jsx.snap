// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthlyTasksRow Unit Tests when normally set up renders correctly 1`] = `
<Fragment>
  <div
    className="bg-orange flex-container-row typical-setup border-bottom"
    data-role="monthlytasks-row"
    style={
      Object {
        "borderLeft": "5px solid white",
      }
    }
  >
    <Connect(InputUncontrolled)
      actions={Object {}}
      inputClassNames="flex-3"
      inputName="content"
      record={Immutable.Map {}}
    />
    <i
      className="fa fa-fw fa-close width-15"
      onClick={[Function]}
      role="button"
      tabIndex="-1"
    />
    <i
      className="fa fa-fw fa-check width-15"
      onClick={[Function]}
      role="button"
      tabIndex="-1"
    />
    <i
      className="fa fa-fw fa-gift width-15"
      onClick={[Function]}
      role="button"
      tabIndex="-1"
    />
  </div>
</Fragment>
`;
